### PR TITLE
update train_text_encoder

### PR DESCRIPTION
**Pull request by dheerajsreddy**
added focal loss and average_precision_score

### CODE REVIEW

File: src/train_text_encoder.py [hunk 1]
```diff
@@ -17,20 +17,38 @@
...
     if torch.cuda.is_available():
         return torch.device("cuda")
+    elif torch.backends.mps.is_available() and torch.backends.mps.is_built():
+        return torch.device("mps")
     else:
         return torch.device("cpu")
 
 
+# @torch.no_grad()
+# def evaluate_auc(model, loader, device):
```
Comment (navneetraju) [inline src/train_text_encoder.py]: Remove commented block. Thats why git exists that we can rollback to commits

File: src/train_text_encoder.py [hunk 2]
```diff
@@ -64,7 +84,8 @@ def run_single_train(
 
     model = AutoModelForSequenceClassification.from_pretrained(model_id, num_labels=2).to(device)
     optimizer = torch.optim.AdamW(model.parameters(), lr=lr, weight_decay=weight_decay)
-    early_stopping = EarlyStopping(patience=patience, verbose=True, mode='max')
+    early_stopping = EarlyStopping(patience=patience, verbose=True)
```
Comment (navneetraju) [inline src/train_text_encoder.py]: set mode to max

File: src/train_text_encoder.py [hunk 3]
```diff
@@ -76,23 +97,27 @@ def run_single_train(
         total_loss = 0.0
         for batch in tqdm(train_loader, desc=f"Training Epoch {epoch}/{epochs}"):
             batch = {k: v.to(device) for k, v in batch.items()}
-            out = model(**batch)
-            loss = out.loss
+            # out = model(**batch)
```
Comment (navneetraju) [inline src/train_text_encoder.py]: remove commented code

File: src/train_text_encoder.py [hunk 4]
```diff
@@ -76,23 +97,27 @@ def run_single_train(
...
+            loss = loss_fn(logits, batch["labels"])#FC
             optimizer.zero_grad()
             loss.backward()
             optimizer.step()
             total_loss += loss.item()
         avg_loss = total_loss / len(train_loader)
 
         val_auc = evaluate_auc(model, val_loader, device)
-        print(f"Epoch {epoch}: Train Loss={avg_loss:.4f}, Val ROC AUC={val_auc:.4f}")
+        # print(f"Epoch {epoch}: Train Loss={avg_loss:.4f}, Val ROC AUC={val_auc:.4f}")
```
Comment (navneetraju) [inline src/train_text_encoder.py]: remove commented code

File: src/train_text_encoder.py [hunk 5]
```diff
@@ -76,23 +97,27 @@ def run_single_train(
...
+        print(f"Epoch {epoch}: Train Loss={avg_loss:.4f}, Val PR AUC={val_auc:.4f}")
 
         if writer:
             writer.add_scalar("Train/Loss", avg_loss, epoch)
-            writer.add_scalar("Val/ROC_AUC", val_auc, epoch)
+            # writer.add_scalar("Val/ROC_AUC", val_auc, epoch)
```
Comment (navneetraju) [inline src/train_text_encoder.py]: remove commented code

File: src/train_text_encoder.py [hunk 6]
```diff
@@ -76,23 +97,27 @@ def run_single_train(
...
+            writer.add_scalar("Val/PR_AUC", val_auc, epoch)
 
         # Early stopping on ROC AUC (maximize)
-        early_stopping(val_auc, model)
+        early_stopping(-val_auc, model)
```
Comment (navneetraju) [inline src/train_text_encoder.py:100]: remove the - once you set mode to 'max' in the early stopping class initialization

File: src/train_text_encoder.py [hunk 7]
```diff
@@ -134,7 +161,9 @@ def train_fn_tune(config, model_id, train_ds, val_ds, epochs: int = 5):
         model.train()
         for batch in train_loader:
             batch = {k: v.to(device) for k, v in batch.items()}
-            loss = model(**batch).loss
+            # loss = model(**batch).loss
```
Comment (navneetraju) [inline src/train_text_encoder.py]: same comments as previous function

File: src/train_text_encoder.py [hunk 8]
```diff
@@ -181,17 +212,19 @@ def main(
         patience: int = typer.Option(3),
         tune_samples: int = typer.Option(5),
         tune_epochs: int = typer.Option(3),
+        alpha: float = typer.Option(1.0, help="Focal loss alpha"),
+        gamma: float = typer.Option(2.0, help="Focal loss gamma"),
         output_dir: str = typer.Option("./text_model"),
 ):
-    full_train = AstroturfTextDataset(os.path.join(dataset_root, "train", "graphs"), model_id)
+    full_train = AstroturfTextDataset(os.path.join(dataset_root, "train"), model_id)
```
Comment (navneetraju) [inline src/train_text_encoder.py]: it should be os.path.join(dataset_root, "train", "graphs") check the GCS bucket structure

File: src/train_text_encoder.py [hunk 9]
```diff
@@ -224,10 +259,15 @@ def main(
             patience=patience,
             save_model=True,
             model_out=output_dir,
-            write_tb=False
+            write_tb=False,
+            alpha=alpha,
+            gamma=gamma
         )
-        print(f"[NO TUNE] Test ROC AUC: {final_auc:.4f}")
+        # print(f"[NO TUNE] Test ROC AUC: {final_auc:.4f}")
```
Comment (navneetraju) [inline src/train_text_encoder.py]: remove comments

File: src/train_text_encoder.py [hunk 10]
```diff
@@ -224,10 +259,15 @@ def main(
...
+        print(f"[NO TUNE] Test PR AUC: {final_auc:.4f}")
 
 
 if __name__ == "__main__":
     app()
+# conda activate swarm_guard
```
Comment (navneetraju) [inline src/train_text_encoder.py]: remove comments

File: src/train_text_encoder.py [hunk 11]
```diff
@@ -148,14 +177,16 @@ def hyperparam_search(model_id, train_ds, val_ds, num_samples: int, max_epochs:
...
+        "gamma": tune.uniform(0.0, 5.0)
     }
     sched = ASHAScheduler(metric="val_roc_auc", mode="max", max_t=max_epochs, grace_period=1)
     analysis = tune.run(
         tune.with_parameters(train_fn_tune, model_id=model_id, train_ds=train_ds, val_ds=val_ds),
         config=space,
         num_samples=num_samples,
         scheduler=sched,
-        resources_per_trial={"cpu": 1, "gpu": 1},
+        resources_per_trial={"cpu": 1, "gpu": 0},
```
Comment (navneetraju) [inline src/train_text_encoder.py]: set to 1, else you wont use GPU

**Review by navneetraju — CHANGES_REQUESTED** [review]

File: src/train_text_encoder.py [hunk 12]
```diff
@@ -17,20 +17,38 @@
     AutoModelForSequenceClassification,
     DataCollatorWithPadding
 )
+from modules.loss.focal_loss import FocalLoss
 
-from src.dataset.astroturf_text_dataset import AstroturfTextDataset
-from src.helpers.early_stopping import EarlyStopping
+from dataset.astroturf_text_dataset import AstroturfTextDataset
```
Comment (navneetraju) [inline src/train_text_encoder.py]: from src.

File: src/train_text_encoder.py [hunk 13]
```diff
@@ -17,20 +17,38 @@
     AutoModelForSequenceClassification,
     DataCollatorWithPadding
 )
+from modules.loss.focal_loss import FocalLoss
```
Comment (navneetraju) [inline src/train_text_encoder.py]: from src.

File: src/train_text_encoder.py [hunk 14]
```diff
@@ -17,20 +17,38 @@
...
 
-from src.dataset.astroturf_text_dataset import AstroturfTextDataset
-from src.helpers.early_stopping import EarlyStopping
+from dataset.astroturf_text_dataset import AstroturfTextDataset
+from helpers.early_stopping import EarlyStopping
```
Comment (navneetraju) [inline src/train_text_encoder.py]: from src.

File: src/train_text_encoder.py [hunk 15]
```diff
@@ -76,36 +82,33 @@ def run_single_train(
         total_loss = 0.0
         for batch in tqdm(train_loader, desc=f"Training Epoch {epoch}/{epochs}"):
             batch = {k: v.to(device) for k, v in batch.items()}
-            out = model(**batch)
-            loss = out.loss
+            logits = model(**batch).logits
```
Comment (navneetraju) [inline src/train_text_encoder.py:85]: im curious why do we need to do .logits ? Isnt it anyways giving the logits ? from what i know it usually dosent give back probs anyways

File: src/train_text_encoder.py [hunk 16]
```diff
@@ -134,7 +138,8 @@ def train_fn_tune(config, model_id, train_ds, val_ds, epochs: int = 5):
         model.train()
         for batch in train_loader:
             batch = {k: v.to(device) for k, v in batch.items()}
-            loss = model(**batch).loss
+            logits = model(**batch).logits
```
Comment (navneetraju) [inline src/train_text_encoder.py:141]: Same question
